src/main/scala/io/iohk/ethereum/ledger/BlockImport.scala

```diff
@@ -96,35 +99,51 @@ class BlockImport(
     handleBlockValidationError(error, block)
   }
 
-  private def handleBlockValidationError(error: ValidationBeforeExecError, block: Block): BlockImportResult = error match {
-    case ValidationBeforeExecError(HeaderParentNotFoundError) =>
-      log.debug(s"Block(${block.idTag}) has unknown parent")
-      UnknownParent
+  private def handleBlockValidationError(error: ValidationBeforeExecError, block: Block): BlockImportResult =
+    error match {
+      case ValidationBeforeExecError(HeaderParentNotFoundError) =>
+        log.debug(s"Block(${block.idTag}) has unknown parent")
+        UnknownParent
 
-    case ValidationBeforeExecError(reason) =>
-      log.debug(s"Block(${block.idTag}) failed pre-import validation")
-      BlockImportFailed(reason.toString)
-  }
+      case ValidationBeforeExecError(reason) =>
+        log.debug(s"Block(${block.idTag}) failed pre-import validation")
+        BlockImportFailed(reason.toString)
+    }
 
   private[ledger] def reorganise(
-    block: Block,
-    currentBestBlock: Block,
-    currentTd: BigInt
+      block: Block,
+      currentBestBlock: Block,
+      currentTd: BigInt
   )(implicit blockExecutionContext: ExecutionContext): Future[BlockImportResult] = Future {
-    blockValidation.validateBlockBeforeExecution(block).fold(error => handleBlockValidationError(error, block), _ => {
-      blockQueue.enqueueBlock(block, currentBestBlock.header.number) match {
-        case Some(Leaf(leafHash, leafTd)) if isBetterBranch(block, currentBestBlock, leafTd, currentTd) =>
-          log.debug("Found a better chain, about to reorganise")
-          reorganiseChain(leafHash, leafTd)
-
-        case _ =>
-          BlockEnqueued
-      }
-    })
+    blockValidation
+      .validateBlockBeforeExecution(block)
+      .fold(
+        error => handleBlockValidationError(error, block),
+        _ => {
+          blockQueue.enqueueBlock(block, currentBestBlock.header.number) match {
+            case Some(Leaf(leafHash, leafTd)) if isBetterBranch(block, currentBestBlock, leafTd, currentTd) =>
+              log.debug("Found a better chain, about to reorganise")
+              reorganiseChain(leafHash, leafTd)
+
+            case _ =>
+              BlockEnqueued
+          }
+        }
+      )
   }
 
-  private def isBetterBranch(block: Block, bestBlock: Block, newTd: BigInt, currentTd: BigInt): Boolean =
-    newTd > currentTd || (blockchainConfig.gasTieBreaker && newTd == currentTd && block.header.gasUsed > bestBlock.header.gasUsed)
+  private def isBetterBranch(block: Block, bestBlock: Block, newTd: BigInt, currentTd: BigInt): Boolean = {
+    lazy val betterTd = newTd > currentTd
+    lazy val tieBreaker =
+      blockchainConfig.gasTieBreaker && newTd == currentTd && block.header.gasUsed > bestBlock.header.gasUsed
+
+    (block.hasCheckpoint, bestBlock.hasCheckpoint) match {
+      case (true, true) => false
```

Comment:
But this way we are opening the door to reverting checkpoints. I'd rather not do that until we have a clear plan on how to deal with that.

Reply:
Hmm and probably it can never be the case that we are comparing blocks from the same branch? That is C5 and C10 from B1<-...<-C5<-...<-C10

Should we at least add a comment then?

I also find it odd that we are allowing reverting checkpoints in all our codebase (i.e. blockchain object and https://github.com/input-output-hk/mantis/pull/728/files#diff-be955b26fb84ed57e0ea600e45c78fb94d45dad720746529236701d67b14e843R49) but here

Reply:
True, there is a door for a revert. What if we changed [this expression](https://github.com/input-output-hk/mantis/blob/f0416e09f629d15fb5454ba7a8cfaaeba82a422c/src/main/scala/io/iohk/ethereum/ledger/BranchResolution.scala#L55) to something like:

```
newCheckpoint.number > oldCheckpoint.number && newBranch.contains(oldCheckpoint)
```

and compare checkpoints by hashes (rather than by numbers) a few lines above.

WDYT?

Reply:
Sounds good to me

However wasn't a rule like this what originally caused the issue on our testnet some time ago? Was that fixed on the OBFT side?